File: programs-phil/spad/vvadd_simd/trillium/splitter.hs
#!/usr/bin/env stack
{- stack --resolver lts-15.2 script 
 --package interpolatedstring-perl6
 --package shakespeare
 --package neat-interpolation
 --package interpolate
 --package text
 --package containers
 --package regex-tdfa
 --package mtl
 --package transformers
 --package split
 --package prettyprinter
-}

--{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE QuasiQuotes #-}
{-# LANGUAGE PatternGuards #-}
import Text.InterpolatedString.Perl6 (qq)
-- import Data.String.Here (i, iTrim, here)
import NeatInterpolation (text)
-- import Text.Shakespeare.Text (st, sbt, text)
-- import Data.String.Interpolate (i)
-- import Data.String.Interpolate.Util (unindent)
import qualified Data.Text as T 
import qualified Data.Text.IO as T_IO
import qualified Data.Map.Strict as M
import Text.Regex.TDFA
import Text.Regex.TDFA.Text ()
import Text.Printf (printf)
import Control.Monad.State.Lazy
import Control.Monad.Trans.Class
import Text.Printf (printf)
import Data.Text.Prettyprint.Doc

data ProgAST = 
  VecSimcV1 {
    kernel_name :: String,
    kernel_body :: [String],
    vissue_blocks :: M.Map Int [String]
  }

emptyProg = 
  VecSimcV1 {
    kernel_name = "",
    kernel_body = [],
    vissue_blocks = M.empty
  }

main :: IO ()
main = splitter "vvadd.c"
-- main = tokenize "vvadd.c"

tokenize :: FilePath -> IO()
tokenize filename = do
  vec_simcv2_src <- readFile filename
  let vec_simcv1_toks = map (show . linewise_tokenize) . lines $ vec_simcv2_src
  sequence_ . map putStrLn $ vec_simcv1_toks

splitter :: FilePath -> IO ()
splitter filename = do
  vec_simcv2_src <- readFile filename
  let vec_simcv1_src = compile vec_simcv2_src
  putStrLn vec_simcv1_src
  where compile = emit_intermingled_vec_simcv1 . linewise_split

emit_intermingled_vec_simcv1 :: ProgAST -> String
emit_intermingled_vec_simcv1 prog = show $
    fun_def "void" (kernel_name prog) ([("int", "trillium_codegen_mask")]++default_args)
      $ vsep
      [(codegen_wrapper "Kernel Start Boilerplate" beginning_boilerplate),
      (codegen_wrapper "Kernel Body" (vsep . map pretty . kernel_body $ prog)),
      (codegen_wrapper "Kernel End Boilerplate" end_boilerplate)]
    where
      default_args = []
  
      fun_def ret_type fun_name params body =
        pretty ret_type <+> pretty fun_name <> pretty_params params <> (nest 2 . braces) (line <> body)
        where pretty_params = tupled . map (\(typ, x) -> pretty typ <+> pretty x)

      fun_call fun_name args = pretty fun_name <> (tupled . map pretty) args

      stmt d = d <> pretty ";"
      stmts = vsep . map stmt

      asm :: [String] -> String -> Doc ann
      asm quals ass = pretty "asm" <+> (hsep . map pretty) quals <> (parens $ pretty ass)

      intermingle scalar vector = vsep
        [pretty "#ifdef SCALAR",
         scalar,
         pretty "#elif defined VECTOR",
         vector,
         pretty "#endif"]

      beginning_boilerplate = 
        intermingle
          (stmt $ fun_call "VECTOR_EPOCH" ["mask"])
          (pretty "//nothing")

      end_boilerplate = 
        intermingle
          (stmts [fun_call "DEVEC" ["devec_0"],
                  asm ["volatile"] "fence",
                  asm [] "scalar return", 
                  pretty "return"])

          (stmts [asm [] "vector return",
                  pretty "return"])

      codegen_wrapper :: String -> Doc ann -> Doc ann
      codegen_wrapper blk_name code_blk = vsep
        [pretty $ ([qq|//***** TRILLIUM CODEGEN: {blk_name} *****|] :: String),
        code_blk,
        pretty [text|//********* TRILLIUM CODEGEN *********|]]

       

    



data VecSimcV2Parser = State VecSimcV2ParserState ProgAST
data VecSimcV2ParserState =
  V2 {fsm_state :: FSMState,
      line_no :: Int,
      prog :: ProgAST,
      curr_vissue_key :: Int}

-- input: Vector-SIMCv2 source text
-- output: Vector-SIMCv1 ast
linewise_split :: String -> ProgAST
linewise_split vec_simcv2_src =
  let final_parser_state = execState (parser vec_simcv2_src) init
  in prog final_parser_state
  where parser :: String -> State VecSimcV2ParserState ()
        parser = foldl (>>) (pure ()) . map linewise_parse . map linewise_tokenize . lines
        init = V2 { fsm_state = FSMKernelBegin,
                    prog = emptyProg,
                    curr_vissue_key = 0,
                    line_no = 0}


-- source formatted as list of code lines, each with a line num.
linewise_parse :: VecSimcV2Line -> State VecSimcV2ParserState ()
linewise_parse line = do
  curr_state <- gets fsm_state
  curr_line_no <- inc_line_no
  case transition curr_state line of
    (FSMError,_) -> return . error . show . vsep . map pretty $
                                ["state transition failure:",
                                printf "attempted transition from state %s" (show curr_state),
                                printf "on line $curr_line_no:",
                                show line]

    (new_state, side_effect) -> do modify (\s -> s {fsm_state = new_state}); side_effect
  where inc_line_no = mk_fresh line_no (\n ps -> ps {line_no= n+1})


  

type ParserEffect = State VecSimcV2ParserState ()
data FSMState = FSMKernelBegin | FSMKernelBody | FSMScalarLoop | FSMVector | FSMKernelEnd | FSMError deriving (Show)
data VecSimcV2Line = KernelBeginPragma String | KernelEndPragma | VectorBeginPragma | VectorEndPragma
                     | ScalarLoopPragma | KernelCode String deriving (Show, Eq)

transition :: FSMState -> VecSimcV2Line -> (FSMState, ParserEffect)
-- copy-paste lines before begin pragma
transition FSMKernelBegin (KernelCode line) =
  (FSMKernelBegin, modify $ insertIntoKernelBody line)
  
transition FSMKernelBegin (KernelBeginPragma name) =
  (FSMKernelBody, modify $ \s -> s {prog = (prog s) { kernel_name=name} } )
transition FSMKernelBody (KernelCode line) = 
  (FSMKernelBegin, modify $ insertIntoKernelBody line)

-- vissue block parse states
transition FSMKernelBody VectorBeginPragma = 
  (FSMVector, 
    do  vissue_key <- mk_fresh_vissue_key
        modify $ \s -> s {  curr_vissue_key = vissue_key })
  where mk_fresh_vissue_key = mk_fresh curr_vissue_key (\n ps -> ps {curr_vissue_key= n+1} )

transition FSMVector (KernelCode line) =
  (FSMVector, modify $ insertIntoVissueBlk line)
        
transition FSMVector VectorEndPragma = 
  (FSMKernelBody, pure ())


-- scalar loop parse states
transition FSMKernelBody ScalarLoopPragma = (FSMScalarLoop, pure ())
transition FSMScalarLoop (KernelCode _) = (FSMKernelBody, pure ())

-- kernel end
transition FSMKernelBody KernelEndPragma = (FSMKernelEnd, pure ())

-- otherwise
transition _ _ = (FSMError, pure ())




linewise_tokenize :: String -> VecSimcV2Line
linewise_tokenize line
  | [name] <- regex_parse line [qq|#pragma trillium vec_simd begin ([{alpha}]+[{alpha_num}|_]*)|]
    = KernelBeginPragma name
  | line =~ "#pragma trillium vec_simd end"
    = KernelEndPragma
  | line =~ "#pragma trillium vector begin"
    = VectorBeginPragma
  | line =~ "#pragma trillium vector end"
    = VectorEndPragma
  | line =~ "#pragma trillium scalar loop"
    = ScalarLoopPragma
  | otherwise
    = KernelCode line
  where 
    alpha = "a-z|A-Z"
    num = "0-9"
    alpha_num = alpha ++ "|" ++ num



regex_parse :: String -> String -> [String]
regex_parse line regex =  
  let (_,_,_,matches) = line =~ regex :: (String,String,String,[String])
  in matches


-- (+++) :: T.Text -> T.Text -> T.Text
-- t1 +++ t2 = T.intercalate "" [t1, t2]


mk_fresh :: (VecSimcV2ParserState -> a)
          -> (a -> VecSimcV2ParserState -> VecSimcV2ParserState)
          -> State VecSimcV2ParserState a
mk_fresh field_getter field_setter = do
  field <- gets field_getter
  modify (field_setter field)
  gets field_getter

-- TODO: inefficient to append to end of list
--        an alternative uses Sequence instead, which allows strict, both-ended appends
-- TODO: Haskell's record field update syntax is too verbose. Learn about and implement a version using Lenses.
insertIntoVissueBlk :: String -> VecSimcV2ParserState -> VecSimcV2ParserState
insertIntoVissueBlk line ps = 
  ps {prog = (prog ps) {vissue_blocks = M.insertWith (flip (++)) (curr_vissue_key ps) [line] (vissue_blocks $ prog ps)}}

insertIntoKernelBody :: String -> VecSimcV2ParserState -> VecSimcV2ParserState
insertIntoKernelBody line s = s {prog = (prog s) { kernel_body= (kernel_body $ prog s) ++ [line]}}

File: programs-phil/spad/vvadd_simd/trillium/splitter.py
import re
import quasiquotes
from enum import Enum, auto

# utility to emit C code
class C_Emitter():
    @staticmethod
    def asm(instr, volatile=False):
        return "asm{}({})".format(" volatile" if volatile else "", instr)

    @staticmethod
    def declare(typed_var):
        qualifiers, type, var = typed_var
        qualifier_string = " ".join(qualifiers) + " " if qualifiers else ""
        return "{}{} {}".format(
                    qualifier_string,
                    type,
                    var)

    @staticmethod
    def fun_sig(ret_type, fun_name, typed_args):
        arg_list = ",".join(list(map(C_Emitter.declare, typed_args)))
        return "{} {}({})".format(ret_type, fun_name, arg_list)


    @staticmethod
    def fun(sig, body):
        print(C_Emitter.to_string(C_Emitter.indent(body)))
        return "{} {{\n{}\n}}".format(sig, C_Emitter.to_string(C_Emitter.indent(body)))

    @staticmethod
    def to_string(code):
        return "\n".join([line + ";" for line in code])

    @staticmethod
    def indent(code, n=1):
        return [" "*(2*n) + line for line in code]


# utility to emit Vector_SIMCv1 code
class Vector_SIMCv1_Emitter():
    blackhole_num = 0
    vissue_key_num = 0

    class ReservedKeywords(Enum):
        MASK = "trillium_codegen_mask"
        DEVEC_ARG = "devec_0"
        BLACKHOLE_VAR = "trillium_codegen_bh_"
        VISSUE_KEY = "trillium_codegen_vissue_key_"

    @staticmethod
    def kernel_sig(kernel_name):
        return "void {}(int {})".format(kernel_name, Vector_SIMCv1_Emitter.ReservedKeywords.MASK.value)

    @staticmethod
    def VECTOR_EPOCH():
        return "VECTOR_EPOCH({})".format(Vector_SIMCv1_Emitter.ReservedKeywords.MASK.value)

    @staticmethod
    def DEVEC():
        return "DEVEC({})".format(Vector_SIMCv1_Emitter.ReservedKeywords.DEVEC_ARG.value)

    @staticmethod
    def boilerplate_intro(vector_variables):
        return Vector_SIMCv1_Emitter.intermingled(
            scalar=[Vector_SIMCv1_Emitter.VECTOR_EPOCH()],
            vector=list(map(C_Emitter.declare,vector_variables)))

    @staticmethod
    def boilerplate_outro():
        return Vector_SIMCv1_Emitter.intermingled(
            scalar=
                [Vector_SIMCv1_Emitter.DEVEC(),
                 C_Emitter.asm("fence", volatile=True),
                 C_Emitter.asm("scalar return")],
            vector=
                [C_Emitter.asm("vector return"),
                 "return"])

    @staticmethod
    def intermingled(scalar=[], vector=[]):
        return ("#ifdef SCALAR\n" + C_Emitter.to_string(scalar) +
               "\n#elif defined VECTOR\n" + C_Emitter.to_string(vector))

    @staticmethod
    def blackhole_var():
        blackhole_num += 1
        return BLACKHOLE_VAR + string(blackhole_num)

    @staticmethod
    def intermingled_emit(kernel_name,
                          vector_variables,
                          vec_simcv1_kernel,
                          vissue_blocks):
        kernel_sig = C_Emitter.fun_sig(
                "void",
                kernel_name,
                [([], "int", Vector_SIMCv1_Emitter.ReservedKeywords.MASK.value)])

        return C_Emitter.fun(kernel_sig,
                [Vector_SIMCv1_Emitter.boilerplate_intro(vector_variables),
                 Vector_SIMCv1_Emitter.boilerplate_outro()])




#utility to parse trillium pragmas
class PragmaKind(Enum):
    VEC_SIMD_START = "#pragma trillium vec_simd (\S*) (\d*) start"
    VEC_SIMD_END = "#pragma trillium vec_simd end"
    VEC_BLK_BEGIN = "#pragma trillium vector begin"
    VEC_BLK_END = "#pragma trillium vector end"

def pragma_parse(pragma_kind, line):
    regex = re.compile(pragma_kind.value)
    match = regex.match(line)
    return match.groups() if match else None


class SplitterState(Enum):
    START = auto()
    KERNEL = auto()
    VEC_BLK = auto()

def split(vec_simcv2):
    # kernel start pragma info: 
    # contains kernel name and core count
    kernel_start_pragma_match = None
    # variables to declare in vector
    vector_variables = []
    # intermingled vec_simcv1 kernel 
    vec_simcv1_kernel = []
    # vissue key/block pairs
    vissue_blocks = {}
    # kernel end pragma
    kernel_end_pragma_match = None

    state = SplitterState.START
    for (line_no, line) in preprocess(vec_simcv2):
        if state == SplitterState.START:
            kernel_start_pragma_match = pragma_parse(PragmaKind.VEC_SIMD_START, line)
            if kernel_start_pragma_match != None:
                print("found kernel start")
                state = SplitterState.KERNEL

        elif state == SplitterState.KERNEL:
            # return after reading exactly one kernel
            # can extend later to read many kernels in a file
            kernel_end_pragma_match = pragma_parse(PragmaKind.VEC_SIMD_END, line)
            if kernel_end_pragma_match != None:
                print("found kernel end")
                break

    if kernel_start_pragma_match == None:
        raise(Exception("did not find a pragma indicating the start of a kernel"))
    elif kernel_end_pragma_match == None:
        raise(Exception("did not find a pragma indicating the end of a kernel"))
    else:
        kernel_name, core_count = kernel_start_pragma_match
        return Vector_SIMCv1_Emitter.intermingled_emit(kernel_name, vector_variables, vec_simcv1_kernel, vissue_blocks)



# utility for preprocessing Vector-SIMCv2 code
def preprocess(vec_simcv2_source):
    return list(enumerate(vec_simcv2_source))

def pretty(vec_simcv1_code):
    return "\n".join(vec_simcv1_code)

if __name__ == "__main__":
    vec_simcv2_file = open("test.c", "r")
    vec_simcv1_file= open("splitter_out.c", "w+")

    vec_simcv2_code = vec_simcv2_file.readlines()

    vec_simcv1_code = split(vec_simcv2_code)

    vec_simcv1_file.write(vec_simcv1_code)
    print("Vector-SIMCv1 code emitted:")
    print(vec_simcv1_code)

    print("Finished.")
